<a href="https://colab.research.google.com/github/Mr-Premanand/Machine-Learning---Supervised-learning-models/blob/main/Regularization_of_LR_Lasso_and_Ridge_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this we have seen that in linear regression we do not get accuracy so much and also we have seen that bias and variance ,from them we can say that model is underfitting(i.e higher bias) and Overfitting(i.e higher variance) so to make model optimum , we perform regularization technique (i.e Lasso and Ridge, details and in https://colab.research.google.com/drive/1WADBzzDo9zCtC-sqXt69PlRJZvQFz4dH?usp=sharing#scrollTo=uTjZT8A76OJb ) so to reduce that error we will not implement above two model and check the acurracy, 

In [ ]:
# Import Libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import datetime as dt
import numpy as np
from sklearn.metrics import r2_score

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Import Database
df=pd.read_csv("/content/drive/MyDrive/Machine learning/Data/covid_19_india.csv", parse_dates=["Date"], dayfirst=True)

In [ ]:
df.head()

Sno       Date     Time State/UnionTerritory ConfirmedIndianNational  \
0    1 2020-01-30  6:00 PM               Kerala                       1   
1    2 2020-01-31  6:00 PM               Kerala                       1   
2    3 2020-02-01  6:00 PM               Kerala                       2   
3    4 2020-02-02  6:00 PM               Kerala                       3   
4    5 2020-02-03  6:00 PM               Kerala                       3   

  ConfirmedForeignNational  Cured  Deaths  Confirmed  
0                        0      0       0          1  
1                        0      0       0          1  
2                        0      0       0          2  
3                        0      0       0          3  
4                        0      0       0          3

In [ ]:
# In we do not need all coluns so that we will keep only required column and again define as df
df=df[['Date','State/UnionTerritory','Cured','Deaths','Confirmed']]

# Change the names of column
df.columns=['date','state','cured','death','confirmed']

In [ ]:
# eirlier record
df.head()

date   state  cured  death  confirmed
0 2020-01-30  Kerala      0      0          1
1 2020-01-31  Kerala      0      0          1
2 2020-02-01  Kerala      0      0          2
3 2020-02-02  Kerala      0      0          3
4 2020-02-03  Kerala      0      0          3

In [ ]:
# latest record
df.tail()

date          state    cured  death  confirmed
18105 2021-08-11      Telangana   638410   3831     650353
18106 2021-08-11        Tripura    77811    773      80660
18107 2021-08-11    Uttarakhand   334650   7368     342462
18108 2021-08-11  Uttar Pradesh  1685492  22775    1708812
18109 2021-08-11    West Bengal  1506532  18252    1534999

In [ ]:
# Now check the cases and death only in maharashtra because of max cases in maharashtra
mh=df[df.state=="Maharashtra"]
mh

date        state    cured   death  confirmed
76    2020-03-09  Maharashtra        0       0          2
91    2020-03-10  Maharashtra        0       0          5
97    2020-03-11  Maharashtra        0       0          2
120   2020-03-12  Maharashtra        0       0         11
133   2020-03-13  Maharashtra        0       0         14
...          ...          ...      ...     ...        ...
17950 2021-08-07  Maharashtra  6130137  133717    6341759
17986 2021-08-08  Maharashtra  6139493  133845    6347820
18022 2021-08-09  Maharashtra  6144388  133996    6353328
18058 2021-08-10  Maharashtra  6151956  134064    6357833
18094 2021-08-11  Maharashtra  6159676  134201    6363442

[520 rows x 5 columns]

# Linear regression in Maharashtra

y=mx+c

In this we will take "Date" as independent variable i.e x
and "Confirmed"cases takes as dependent variable i.e y  ,   so that by puting the date we can easily predict the cases.

In [ ]:
# Converting date to ordinal
mh["date"]=mh["date"].map(dt.datetime.toordinal)
mh.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


date        state  cured  death  confirmed
76   737493  Maharashtra      0      0          2
91   737494  Maharashtra      0      0          5
97   737495  Maharashtra      0      0          2
120  737496  Maharashtra      0      0         11
133  737497  Maharashtra      0      0         14

In [ ]:
# Define dependent and independent variables
x=mh[['date']]
y=mh[['confirmed']]

Step-3

Split the date into train and test set

In [ ]:
# here we have given test size 0.3 i.e 30% data from dataset will be consider as test data
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3, random_state = 0)      #when random_state set to an integer, train_test_split will return same results for each execution

Step-4
fit the data into linear regresson model

In [ ]:
# we have imported linearRegression Lib already above so we define it as notaion lr2 for Model-2
lr=LinearRegression()

In [ ]:
# Fit the train data
lr.fit(x_train,y_train)

LinearRegression()

Step-5

Now feed the test data to test the accuracy i.e feed the x_test data 

In [ ]:
# now we will test data over by using independet variable x_test to check result as per the y_test
# use above trained linear regression relation
y_pred=lr.predict(x_test)
y_pred

array([[ 4233712.90472603],
       [  175330.74784279],
       [ 2056358.66817665],
       [ 1824451.11635399],
       [ 2803616.33515739],
       [-1009974.51702499],
       [  523192.07557487],
       [ 2185196.19696617],
       [ 3757014.04820442],
       [   46493.21905327],
       [ 4375434.18639565],
       [ 4091991.62305832],
       [  780867.13315582],
       [ 4517155.46806526],
       [ 3666827.77805138],
       [ -198298.08564758],
       [ 3009756.38122177],
       [  561843.33421326],
       [ 4298131.66912079],
       [ 2636127.5477314 ],
       [ -726531.95368767],
       [  -43693.05109978],
       [ 3241663.93304443],
       [ 1772916.10483742],
       [ 3357617.70895576],
       [ 1631194.82316971],
       [ -623461.93065453],
       [ 3228780.18016434],
       [ 1373519.76558876],
       [ -932671.99975014],
       [ -610578.17777634],
       [ 2739197.57076263],
       [ 2868035.09955215],
       [ 1051425.94361496],
       [ 5212878.12352943],
       [  110911.983

In [ ]:
#Now vie actual y_test values to check wheather y_pred values are come as per y_test
y_test.head()

confirmed
14566    4822902
3370      139010
8505     1749777
7875     1666668
10570    1978044

Step
Find MSE 

In [ ]:
# we already imported MSE lib so we will directly find the MSE
# mean squre error of model 1
mean_squared_error(y_test,y_pred)

547107733817.5716

In [ ]:
# Also check the accuracy
lr.score(x_test,y_test)

0.8743818261261916

# Now,  Lasso and Ridge model

In [ ]:
# Import the lasso and ridge model from sklearn
from sklearn.linear_model import Lasso, Ridge

Ridge Model

In [ ]:
# Give notaion to Ridge model
rd=Ridge()

#fit the values in Model
rd.fit(x_train,y_train)

y_pred_ridge =rd.predict(x_test)

In [ ]:
#Check the accuracy
acc = rd.score(x_test,y_pred_ridge)
print("Accuracy :", acc )

MSE  = mean_squared_error(y_test,y_pred_ridge)
print("MSE :" , MSE)

RMSE = np.sqrt(MSE)
print("RMSE :" ,RMSE)

r2 = r2_score(y_test,y_pred_ridge)
print("R2 :" ,r2)
print("Adjusted R2 : ",1-(1-r2_score(y_test,y_pred_ridge))*((x_test.shape[0]-1)/(x_test.shape[0]-x_test.shape[1]-1)))

Accuracy : 1.0
MSE : 547107739449.22375
RMSE : 739667.3167372098
R2 : 0.8743818248331412
Adjusted R2 :  0.8735661223969927


As we seen above accuracy of ridge and linear regression is quite similar but last some digits shows that accuracy of ridge model is higher

Now to increase the more accuracy we have to adjust the value of λ,
as belows formula
$$\text{RSS} + \lambda \sum_{j=1}^p \beta_j^2$$

In [ ]:
# Now adjust the value of λ
rd1=Ridge(alpha=3)           # by default it taken as 1

#fit the values in Model
rd1.fit(x_train,y_train)

y_pred_ridge_1=rd1.predict(x_test)

In [ ]:
#Check the accuracy
acc = rd.score(x_test,y_pred_ridge_1)
print("Accuracy :", acc )

MSE  = mean_squared_error(y_test,y_pred_ridge_1)
print("MSE :" , MSE)

RMSE = np.sqrt(MSE)
print("RMSE :" ,RMSE)

r2 = r2_score(y_test,y_pred_ridge_1)
print("R2 :" ,r2)
print("Adjusted R2 : ",1-(1-r2_score(y_test,y_pred_ridge_1))*((x_test.shape[0]-1)/(x_test.shape[0]-x_test.shape[1]-1)))

Accuracy : 0.9999999999999404
MSE : 547107750712.54987
RMSE : 739667.3243509881
R2 : 0.8743818222470355
Adjusted R2 :  0.8735661197940943


AS we seen above as increasing the alpha value accuracy got decrease,

so now we will reduce the alpha value



In [ ]:
# Now adjust the value of λ by reducing
rd2=Ridge(alpha=0.5)           # by default it taken as 1

#fit the values in Model
rd2.fit(x_train,y_train)

#Check the accuracy
rd2.score(x_test,y_test)

0.874381825479638

As seen accuracy start to increase slitly

# Lasso Model

In [ ]:
# Give notaion to Ridge model
ls=Lasso()

#fit the values in Model
ls.fit(x_train,y_train)

y_pred_lasso = ls.predict(x_test)

In [ ]:
#Check the accuracy
acc = rd.score(x_test,y_pred_lasso)
print("Accuracy :", acc )

MSE  = mean_squared_error(y_test,y_pred_lasso)
print("MSE :" , MSE)

RMSE = np.sqrt(MSE)
print("RMSE :" ,RMSE)

r2 = r2_score(y_test,y_pred_lasso)
print("R2 :" ,r2)
print("Adjusted R2 : ",1-(1-r2_score(y_test,y_pred_lasso))*((x_test.shape[0]-1)/(x_test.shape[0]-x_test.shape[1]-1)))

Accuracy : 0.9999999999999859
MSE : 547107733976.8353
RMSE : 739667.3130379869
R2 : 0.874381826089624
Adjusted R2 :  0.8735661236616346


as seen above lasso have slietly higher accuracy than linear regression(0.8743818248331412) and Ridge model(0.8743818248331412)

Now, adjust the alpha values in that

$$\text{RSS} + \lambda \sum_{j=1}^p \lvert \beta_j \rvert$$

In [ ]:
# Give notaion to Ridge model
ls1=Lasso(alpha=2)

#fit the values in Model
ls1.fit(x_train,y_train)

#Check the accuracy
ls1.score(x_test,y_test)        

0.8743818260530626

As seen above as increasing the alpha value its accuracy start to decrese

now we will reduce the alha value in model

In [ ]:
# Give notaion to Ridge model
ls2=Lasso(alpha=0.3)

#fit the values in Model
ls2.fit(x_train,y_train)

y_pred_lasso_2=ls2.predict(x_test)      

In [ ]:
#Check the accuracy
acc = rd.score(x_test,y_pred_lasso_2)
print("Accuracy :", acc )

MSE  = mean_squared_error(y_test,y_pred_lasso_2)
print("MSE :" , MSE)

RMSE = np.sqrt(MSE)
print("RMSE :" ,RMSE)

r2 = r2_score(y_test,y_pred_lasso_2)
print("R2 :" ,r2)
print("Adjusted R2 : ",1-(1-r2_score(y_test,y_pred_lasso_2))*((x_test.shape[0]-1)/(x_test.shape[0]-x_test.shape[1]-1)))

Accuracy : 0.9999999999999853
MSE : 547107733865.41296
RMSE : 739667.3129626677
R2 : 0.8743818261152071
Adjusted R2 :  0.8735661236873837


As we seen that accuracy is increased than above lasso model

So, every time we need to adjust the alpha value to get best fit aplha value, and that value manually impossible, so that to get most accuracte alpha value we have technique name "hiperparameter tunning" which we will learn in next session

# Cross-Validation and Hyperparameter Tunning

In this we will find the best fit model hyperparameter (i.e alpha or lambda) by using CV and GridSearchCV

In [ ]:
# Import GridSearchCV model
from sklearn.model_selection import GridSearchCV

In [ ]:
### Cross validation
lasso = Lasso()
parameters = {'alpha': [1e-15,1e-13,1e-10,1e-8,1e-5,1e-4,1e-3,1e-2,1e-1,1,5,10,20,30,40,45,50,55,60,100]}
lasso_regressor = GridSearchCV(lasso, parameters, scoring='neg_mean_squared_error', cv=3)
lasso_regressor.fit(x_train, y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 6.724e+13, tolerance: 9.983e+10
  coef_, l1_reg, l2_reg, X, y, max_iter, tol, rng, random, positive
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 6.585e+13, tolerance: 1.118e+11
  coef_, l1_reg, l2_reg, X, y, max_iter, tol, rng, random, positive
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_coordinate_descent.py:648: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisa

GridSearchCV(cv=3, estimator=Lasso(),
             param_grid={'alpha': [1e-15, 1e-13, 1e-10, 1e-08, 1e-05, 0.0001,
                                   0.001, 0.01, 0.1, 1, 5, 10, 20, 30, 40, 45,
                                   50, 55, 60, 100]},
             scoring='neg_mean_squared_error')

In [ ]:
print("The best fit alpha value is found out to be :" ,lasso_regressor.best_params_)
print("\nUsing ",lasso_regressor.best_params_, " the negative mean squared error is: ", lasso_regressor.best_score_)

The best fit alpha value is found out to be : {'alpha': 1e-15}

Using  {'alpha': 1e-15}  the negative mean squared error is:  -552067975927.0084


In [ ]:
# fit the data in newl trained model
y_pred_lasso = lasso_regressor.predict(x_test)

In [ ]:
#Check the accuracy
acc = rd.score(x_test,y_pred_lasso)
print("Accuracy :", acc )

MSE  = mean_squared_error(y_test,y_pred_lasso)
print("MSE :" , MSE)

RMSE = np.sqrt(MSE)
print("RMSE :" ,RMSE)

r2 = r2_score(y_test,y_pred_lasso)
print("R2 :" ,r2)
print("Adjusted R2 : ",1-(1-r2_score(y_test,y_pred_lasso))*((x_test.shape[0]-1)/(x_test.shape[0]-x_test.shape[1]-1)))

Accuracy : 0.9999999999999851
MSE : 547107733817.5716
RMSE : 739667.3129303278
R2 : 0.8743818261261916
Adjusted R2 :  0.8735661236984397
